add bypass metric

baberabb opened the pull request:
Started on #1152. Couple of issues:

- [x] Cannot amend `metric_list` in evaluate
- [x] print results crashing
- [x] repeated generations are not saved in `log_samples`

haileyschoelkopf commented on lm_eval/__main__.py:
```diff
@@ -142,6 +142,12 @@ def parse_eval_args() -> argparse.Namespace:
         metavar="CRITICAL|ERROR|WARNING|INFO|DEBUG",
         help="Controls the reported logging error level. Set to DEBUG when testing + adding new task configurations for comprehensive log output.",
     )
+    parser.add_argument(
```
might be worth adding an assert in evaluate() that checks we haven't set predict_only and not log_samples?

StellaAthena replied:
Alternatively, we can just auto-set log_samples if predict_only is used to decrease the cognitive load on the user

baberabb commented:
@haileyschoelkopf ready for review! I called the metric `bypass`. I don't know if you want to keep that or change it to `predict_only`. 

Also had to get a bit hacky in changing the task attributes post-hoc if we are calling from `__main__`. Couldn't find an easier way.

haileyschoelkopf commented on lm_eval/evaluator.py:
```diff
@@ -130,6 +134,17 @@ def simple_evaluate(
+                f"Processing {task_name} in output-only mode. Metrics will not be calculated!"
+            )
+            # we have to change the class properties post-hoc. This is pretty hacky.
+            task_obj._metric_fn_list = {"bypass": lambda x: None}
```
maybe we can abstract this code that accesses private task class attributes into something like `ConfigurableTask.override_metrics()`? And/or an `ConfigurableTask.update_config()` method for updating the config it's based on

haileyschoelkopf reviewed (changes requested):
This looks really great to me, thanks for working on it!!

Left one comment as a suggestion for code cleanup/readability (avoiding accessing private task attributes), if you're willing to address it in this PR. 


When I test this I see the following:

```
accelerate launch -m lm_eval --model hf --tasks truthfulqa_mc1,truthfulqa_mc2 --batch_size auto --predict_only --output_path ./test-bypass
The following values were not passed to `accelerate launch` and had defaults used instead:
        `--num_processes` was set to a value of `1`
        `--num_machines` was set to a value of `1`
        `--mixed_precision` was set to a value of `'no'`
        `--dynamo_backend` was set to a value of `'no'`
To avoid this warning pass in values for each of the problematic parameters or run `accelerate config`.
2023-12-18:21:34:10,665 INFO     [utils.py:148] Note: NumExpr detected 40 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
2023-12-18:21:34:15,738 INFO     [__main__.py:162] Verbosity set to INFO
2023-12-18:21:34:31,091 INFO     [__main__.py:238] Selected Tasks: ['truthfulqa_mc1', 'truthfulqa_mc2']
2023-12-18:21:34:31,097 INFO     [huggingface.py:152] Using device 'cuda'
2023-12-18:21:34:36,048 INFO     [evaluator.py:139] Processing truthfulqa_mc1 in output-only mode. Metrics will not be calculated!
2023-12-18:21:34:36,049 INFO     [evaluator.py:139] Processing truthfulqa_mc2 in output-only mode. Metrics will not be calculated!
2023-12-18:21:34:36,049 INFO     [task.py:356] Building contexts for task on rank 0...
2023-12-18:21:34:37,606 INFO     [task.py:356] Building contexts for task on rank 0...
2023-12-18:21:34:39,097 INFO     [evaluator.py:341] Running loglikelihood requests
  0%|                                               | 0/9996 [00:00<?, ?it/s]Passed argument batch_size = auto:1. Detecting largest batch size
Determined largest batch size: 64
100%|███████████████████████████████████| 9996/9996 [00:58<00:00, 170.94it/s]
Traceback (most recent call last):
  File "/mnt/ssd-1/hailey/miniconda3/envs/harness-dev/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/mnt/ssd-1/hailey/miniconda3/envs/harness-dev/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/mnt/ssd-1/hailey/lm-evaluation-harness/lm_eval/__main__.py", line 291, in <module>
    cli_evaluate()
  File "/mnt/ssd-1/hailey/lm-evaluation-harness/lm_eval/__main__.py", line 240, in cli_evaluate
    results = evaluator.simple_evaluate(
  File "/mnt/ssd-1/hailey/lm-evaluation-harness/lm_eval/utils.py", line 404, in _wrapper
    return fn(*args, **kwargs)
  File "/mnt/ssd-1/hailey/lm-evaluation-harness/lm_eval/evaluator.py", line 167, in simple_evaluate
    results = evaluate(
  File "/mnt/ssd-1/hailey/lm-evaluation-harness/lm_eval/utils.py", line 404, in _wrapper
    return fn(*args, **kwargs)
  File "/mnt/ssd-1/hailey/lm-evaluation-harness/lm_eval/evaluator.py", line 503, in evaluate
    agg_fn = task.aggregation()[metric]
KeyError: 'acc'
```

So I think there might be a few lingering references to the old overridden metrics / aggregation fns / etc. that need to be scrubbed still.

haileyschoelkopf commented:
@lintangsutawika any updated thoughts on this, now that we are not planning to rework the metric / aggregation abstraction any time soon?

I'm quite a big fan of doing things in the way this PR implements it.

lintangsutawika commented:
I'm not entirely on board, but I see this as needed for `--predict_only`

baberabb commented:
ok have tested it out and looks to be in working order. No major nits except `logliklihood` table gets printed like this:

```
hf (pretrained=EleutherAI/pythia-70m,device=mps), gen_kwargs: (None), limit: 20.0, num_fewshot: None, batch_size: 8
|Tasks|Version|Filter|n-shot|Metric|Value|   |Stderr|
|-----|-------|------|------|------|-----|---|------|
```
And if its a mixed run then the logliklihood tasks do not print. The samples are saved fine though. IMO will be used almost exclusively for `generation` tasks anyway.

lintangsutawika commented:
If predict_only is exclusively for generation do we still want to allow users to mix task of different output types and while using the arg? It might be better not to and also then skip the table (skip metric any metric calculation).

baberabb commented:
It works for both tasks but can't think of when people would want to use it for logliklihood. Maybe for debugging or experimenting? I can change it so that only generation scoring is bypassed using the `--predict_only` flag.

lintangsutawika commented:
Bypassing only generation tasks makes sense.

baberabb commented:
Also noticed we aren't passing `do_sample` when people change the gen_kwargs. Should we force that or is it expected people know they need to pass that on as well. There's usually a warning from HF.

haileyschoelkopf commented:
> Bypassing only generation tasks makes sense.

This seems nonintuitive for users. A warning should be given when running non-generate tasks saying they’ll still be scored if we pass predict_only, at least

Also, we’ll need to document this CLI arg and feature in the docs!

baberabb commented:
Updated the warning for `gen_kwargs`. The previous one said "these settings will be used over set parameters in yaml tasks" but we only _updated_ the dict though? Not sure which was the intended functionality. Tried to make the current one more obvious.

For `--predict_only` a specific warning is generated for each task changed.

baberabb commented:
Added come drafts in the `Readme`! Also moved the table to the bottom, thought this big (and increasingly growing!) thing broke the flow of the document. Let me know what if this is better or I should revert it.

haileyschoelkopf commented on lm_eval/evaluator.py:
```diff
@@ -133,21 +133,18 @@ def simple_evaluate(
-            task_obj.override_config(
-                key="generation_kwargs", value=gen_kwargs, update=True
-            )
+        if task_obj.get_config("output_type") == "generate_until":
```
We should add a warning that  `--predict_only` will be ignored for non-generate_until tasks.

baberabb replied:
It works for all. Should I restrict it to generate_until?

haileyschoelkopf replied:
misread the behavior, I think allowing it to work for all is more intuitive, that’s fine!

haileyschoelkopf reviewed (commented):
Quite happy with this PR overall! I'll add some nits. I'd very much like to merge early next week once I've run this again myself and looked at some log files, prior to taking action on #1167 

Let's leave the model_outputs.ipynb and `humaneval` out of the PR for now and do them in a subsequent one (though they are great additions!)

baberabb commented:
Sounds good! Shouldn't this be merged after #1167? Looks like most of the workarounds here won't be needed after that. 

Also thinking about it, not really sold on the `predict_only` flag. Under what circumstances would someone want to use it (as opposed to `bypass` metric in the config)?. Seems like you can just run it normally and get the metric for free if you just want the samples.

haileyschoelkopf commented:
> Shouldn't this be merged after https://github.com/EleutherAI/lm-evaluation-harness/pull/1167? Looks like most of the workarounds here won't be needed after that.

Sure, let's see what our timeline on merging #1167 is next week and go from there.

> Also thinking about it, not really sold on the predict_only flag. Under what circumstances would someone want to use it (as opposed to bypass metric in the config)?. Seems like you can just run it normally and get the metric for free if you just want the samples.

I think it's still valuable to have! For example, in the Llemma sympy-checked math tasks, for Maj@K at high K, doing the scoring actually takes way more wall-clock time than inference. Because it's CPU-bottlenecked and one wants to conserve GPU-hours / minimize time until logging results, it's valuable to do this heavy answer checking offline on a separate machine. However, allowing for it to give metrics at runtime is still valuable for QoL if someone doesn't care about this delay.

Also, the user who originally requested some version of this feature wanted it in order to run on tasks that are typically scored online, but wanted to run some postprocessing outside the library manually then score those edited results as normal.

baberabb commented:
> I think it's still valuable to have! For example, in the Llemma sympy-checked math tasks, for Maj@K at high K, doing the scoring actually takes way more wall-clock time than inference.

aah yeah that makes more sense! I was thinking more along the lines of just calculating acc or exact match.

haileyschoelkopf reviewed (approved):
Since we’re deferring metric overhaul for now in favor of other more urgent features like #1321 , merging this in.

Thanks very much @baberabb !

LSinev commented on lm_eval/evaluator.py:
```diff
@@ -129,25 +133,30 @@ def simple_evaluate(
-            config["generation_kwargs"].update(gen_kwargs)
+        if task_obj.get_config("output_type") == "generate_until":
+            if gen_kwargs is not None:
+                task_obj.override_config(
```
Seems like source for fails, when task is an instance of Task with no override_config available

LSinev commented on lm_eval/evaluator.py:
```diff
@@ -129,25 +133,30 @@ def simple_evaluate(
+                    f"Processing {task_name} in output-only mode. Metrics will not be calculated!"
+                )
+                # we have to change the class properties post-hoc. This is pretty hacky.
+                task_obj.override_metric(metric_name="bypass")
```
Seems like source for fails too, when task is an instance of Task with no override_metric available. Did not find any PR addressing this. May be missing something.

lintangsutawika replied:
@baberabb could you look into this?

baberabb replied:
@LSinev thanks for catching this! I'll just move this to `Task`
